File: api/app/schemas/library.py
"""Library wire schema — request / response models for the Library endpoints.

The Library wire shape mirrors :class:`SectionInstance.data` for entry-based
section kinds. Type-aware data validation lives in ``app.schema`` (the AST
source of truth). HTTP I/O shapes stay here so codegen continues to pick up
the AST models while the service layer can validate incoming payloads
against the closed ``LibraryEntryKind`` enum.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel

from app.schema.models import LibraryEntryKindStr, SectionInstance


class LibraryEntryKind(str, Enum):
    """Closed set of entry-based section kinds eligible for the Library.

    The CV's ``profile`` and ``summary`` sections have dict-shaped
    ``SectionInstance.data`` and are excluded from Library promotion —
    they stay authored directly inside the CV.
    """

    EXPERIENCE = "experience"
    EDUCATION = "education"
    SKILL = "skill"
    PROJECT = "project"
    CERTIFICATION = "certification"
    LANGUAGE = "language"


class LibraryEntryCreate(BaseModel):
    """Request body to create a Library entry."""

    kind: LibraryEntryKind
    payload: list[dict] = []


class LibraryEntryUpdate(BaseModel):
    """Request body to mutate a Library entry's payload."""

    payload: list[dict]


class LibraryEntryResponse(BaseModel):
    """Response shape for a Library entry.

    ``model_config = {"from_attributes": True}`` lets the route return an
    ORM row directly; ``payload`` accepts the JSON column as a Python list.
    """

    model_config = {"from_attributes": True}

    id: str
    kind: LibraryEntryKind
    payload: list[dict]
    created_at: datetime
    updated_at: datetime


class LibraryCloneResponse(BaseModel):
    """Response shape for the clone endpoint.

    ``section_instance`` is a ready-to-paste ``SectionInstance`` that the
    CV builder can append to ``cvStore.currentCV.sections`` directly.
    """

    section_instance: SectionInstance


class PromoteToLibraryResponse(BaseModel):
    """Response shape for promoting a CV's entries into the Library.

    ``promoted`` maps each eligible kind to the number of new entries
    created on this call (excluding content-hash duplicates). ``skipped``
    lists section IDs whose ``type`` is not eligible (e.g. ``profile``)
    so the FE can surface them in a toast.
    """

    library_id: str
    promoted: dict[str, int]
    skipped: list[str]


class AddEntryToLibraryResponse(BaseModel):
    """Response shape for ``POST /cvs/{cv_id}/sections/.../add-to-library``.

    ``created=False`` means an identical Library entry already exists
    (same kind + content hash); the FE should treat this as a no-op
    success. ``entry_id`` is the new or existing Library entry's id.
    """

    library_id: str
    entry_id: str | None
    created: bool


class AddEntryToLibraryRequest(BaseModel):
    """Current CV entry snapshot submitted by the builder.

    The builder uses manual saves, so the visible entry can be newer than
    the copy persisted on the CV. The section kind is included so a newly
    added, not-yet-saved section can still be routed correctly.
    """

    kind: LibraryEntryKind
    entry: dict




__all__ = [
    "LIBRARY_ENTRY_KINDS",
    "LIBRARY_KIND_TO_SECTION_TYPE",
    "SECTION_TYPE_TO_LIBRARY_KIND",
    "AddEntryToLibraryRequest",
    "AddEntryToLibraryResponse",
    "LibraryEntryCreate",
    "LibraryEntryKind",
    "LibraryEntryResponse",
    "LibraryEntryUpdate",
    "LibraryCloneResponse",
    "PromoteToLibraryResponse",
    "library_kind_for_section_type",
    "section_type_for_library_kind",
]


# Tuple form for membership checks (mirrors the Literal in app.schema.models).
LIBRARY_ENTRY_KINDS: tuple[LibraryEntryKindStr, ...] = tuple(
    kind.value for kind in LibraryEntryKind
)

# CV sections use plural names for four entry-based sections while Library
# rows use singular kinds. Keep the translation at this boundary so every
# service operation handles the same vocabulary.
SECTION_TYPE_TO_LIBRARY_KIND: dict[str, LibraryEntryKindStr] = {
    "experience": "experience",
    "education": "education",
    "skills": "skill",
    "projects": "project",
    "languages": "language",
    "certifications": "certification",
    # Accept legacy singular section names while old CV rows are migrated
    # naturally through the next write.
    "skill": "skill",
    "project": "project",
    "language": "language",
    "certification": "certification",
}
LIBRARY_KIND_TO_SECTION_TYPE: dict[str, str] = {
    "experience": "experience",
    "education": "education",
    "skill": "skills",
    "project": "projects",
    "language": "languages",
    "certification": "certifications",
}


def library_kind_for_section_type(section_type: str) -> LibraryEntryKindStr | None:
    return SECTION_TYPE_TO_LIBRARY_KIND.get(section_type)


def section_type_for_library_kind(kind: str) -> str | None:
    return LIBRARY_KIND_TO_SECTION_TYPE.get(kind)

File: web/src/lib/api/library.ts
import client from "./client";

// ─── Library types ──────────────────────────────────────────────────
//
// LibraryEntryKind + LibraryEntryResponse are declared here as plain TS
// types rather than imported from `../generated/schema` because the
// codegen script only scans `app.schema.models`. The HTTP wire shapes
// live in `app.schemas.library` and are not picked up by design
// (per Phase A2 / plan assumption A8).
//
// If/when the codegen script is extended to scan `app.schemas.*`,
// these declarations can be replaced with imports from the generated
// file.

import type { SectionType } from "../sections/types";

export type LibraryEntryKind =
  | "experience"
  | "education"
  | "skill"
  | "project"
  | "certification"
  | "language"
  | "research";

export const LIBRARY_KIND_TO_SECTION_TYPE: Record<LibraryEntryKind, SectionType> = {
  experience: "experience",
  education: "education",
  skill: "skills",
  project: "projects",
  certification: "certifications",
  language: "languages",
  research: "research",
};

export const SECTION_TYPE_TO_LIBRARY_KIND: Partial<Record<SectionType, LibraryEntryKind>> = {
  experience: "experience",
  education: "education",
  skills: "skill",
  projects: "project",
  certifications: "certification",
  languages: "language",
  research: "research",
};

export function sectionTypeForLibraryKind(kind: LibraryEntryKind): SectionType {
  return LIBRARY_KIND_TO_SECTION_TYPE[kind];
}

export function libraryKindForSectionType(sectionType: string): LibraryEntryKind | undefined {
  return SECTION_TYPE_TO_LIBRARY_KIND[sectionType as SectionType];
}

export function isLibraryKind(kind: string): kind is LibraryEntryKind {
  return (LIBRARY_KINDS as readonly string[]).includes(kind);
}

export const LIBRARY_KIND_LABELS: Record<LibraryEntryKind, string> = {
  experience: "Experiences",
  education: "Education",
  skill: "Skills",
  project: "Projects",
  certification: "Certifications",
  language: "Languages",
  research: "Research",
};

export const LIBRARY_KINDS: LibraryEntryKind[] = [
  "experience",
  "education",
  "skill",
  "project",
  "certification",
  "language",
  "research",
];

export interface LibraryEntry {
  id: string;
  kind: LibraryEntryKind;
  payload: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
}

export interface LibraryCloneResponse {
  section_instance: {
    id: string;
    type: SectionType;
    title: string;
    enabled: boolean;
    data: unknown;
    style: unknown;
  };
}

export interface PromoteToLibraryResponse {
  library_id: string;
  promoted: Record<string, number>;
  skipped: string[];
}

// ─── API surface ────────────────────────────────────────────────────

export async function listLibrary(kind?: LibraryEntryKind): Promise<LibraryEntry[]> {
  const { data } = await client.get("/library", { params: kind ? { kind } : {} });
  return data;
}

export async function createLibrary(
  kind: LibraryEntryKind,
  payload: Array<Record<string, unknown>>,
): Promise<LibraryEntry> {
  const { data } = await client.post("/library", { kind, payload });
  return data;
}

export async function updateLibrary(
  id: string,
  payload: Array<Record<string, unknown>>,
): Promise<LibraryEntry> {
  const { data } = await client.patch(`/library/${id}`, { payload });
  return data;
}

export async function deleteLibrary(id: string): Promise<void> {
  await client.delete(`/library/${id}`);
}

export async function cloneLibrary(id: string): Promise<LibraryCloneResponse> {
  const { data } = await client.post(`/library/${id}/clone`);
  return data;
}

export async function promoteCvToLibrary(cvId: string): Promise<PromoteToLibraryResponse> {
  const { data } = await client.post(`/cvs/${cvId}/promote-to-library`);
  return data;
}

export interface AddEntryToLibraryResponse {
  library_id: string;
  entry_id: string | null;
  created: boolean;
}

export interface AddEntryToLibraryData {
  kind: LibraryEntryKind;
  entry: Record<string, unknown>;
}

export async function addEntryToLibrary(
  cvId: string,
  sectionId: string,
  entryId: string,
  payload: AddEntryToLibraryData,
): Promise<AddEntryToLibraryResponse> {
  const { data } = await client.post(
    `/cvs/${cvId}/sections/${sectionId}/entries/${entryId}/add-to-library`,
    payload,
  );
  return data;
}
